sdk/ml/azure-ai-ml/azure/ai/ml/_azure_environments.py

```diff
@@ -189,3 +192,130 @@ def _resource_to_scopes(resource):
+            EndpointURLS.ACTIVE_DIRECTORY_ENDPOINT: cloud["authentication"]["loginEndpoint"],
+            EndpointURLS.AML_RESOURCE_ID: cloud["resourceManager"],
+            EndpointURLS.STORAGE_ENDPOINT: cloud["suffixes"]["storage"]
+        }
+    return cli_cloud_metadata_dict
+
+def _get_cloud(cloud_name):
+    return next((data for name,data in _get_all_clouds().items() if name == cloud_name), None)
+
+
+def _get_all_clouds():
+    # Start with the hard coded list of clouds in this file
+    all_clouds = {}
+    all_clouds.update(_environments)
+    # Get configs from the config file
+    config = configparser.ConfigParser()
+    try:
+        config.read(CLOUD_CONFIG_FILE)
+    except configparser.MissingSectionHeaderError:
+        os.remove(CLOUD_CONFIG_FILE)
+        module_logger.warning("'%s' is in bad format and has been removed.", CLOUD_CONFIG_FILE)
+    for section in config.sections():
+        all_clouds[section] = dict(config.items(section))
+    # Now do the metadata URL
+    arm_url = os.environ[_ARM_METADATA_URL_ENV_NAME] if _ARM_METADATA_URL_ENV_NAME in os.environ else _DEFAULT_ARM_URL
+    all_clouds.update(_get_clouds_by_metadata_url(arm_url))
+    # Send them all along with the hardcoded environments
+    return all_clouds
+
+class CloudNotRegisteredException(Exception):
+    def __init__(self, cloud_name):
+        super(CloudNotRegisteredException, self).__init__(cloud_name)
+        self.cloud_name = cloud_name
+
+    def __str__(self):
+        return "The cloud '{}' is not registered.".format(self.cloud_name)
+
+
+class CloudAlreadyRegisteredException(Exception):
+    def __init__(self, cloud_name):
+        super(CloudAlreadyRegisteredException, self).__init__(cloud_name)
+        self.cloud_name = cloud_name
+
+    def __str__(self):
+        return "The cloud '{}' is already registered.".format(self.cloud_name)
+
+def _config_add_cloud(config, cloud, overwrite=False):
+    """ Add a cloud to a config object """
+    try:
+        config.add_section(cloud["name"])
+    except configparser.DuplicateSectionError:
+        if not overwrite:
+            raise CloudAlreadyRegisteredException(cloud["name"])
+    for k,v in cloud.items():
+        if k != "name" and v is not None:
+            config.set(cloud["name"], k, v)
+
+def _save_cloud(cloud, overwrite=False):
+    config = configparser.ConfigParser()
+    config.read(CLOUD_CONFIG_FILE)
+    _config_add_cloud(config, cloud, overwrite=overwrite)
+    if not os.path.isdir(GLOBAL_CONFIG_DIR):
+        os.makedirs(GLOBAL_CONFIG_DIR)
+    with open(CLOUD_CONFIG_FILE, 'w') as configfile:
+        config.write(configfile)
+
+def _add_cloud(cloud):
+    if _get_cloud(cloud["name"]):
+        raise CloudAlreadyRegisteredException(cloud["name"])
+    _save_cloud(cloud)
+
+
+def _update_cloud(cloud):
+    if not _get_cloud(cloud["name"]):
+        raise CloudNotRegisteredException(cloud["name"])
+    _save_cloud(cloud, overwrite=True)
```

Comment:
do we need any of this? I guess this would be for the user to be able to unblock themselves if they know the cloud registration details.